### DISCUSSION TITLE
quorum queues go to noproc on restarting of nodes. New issue with 3.13

### DISCUSSION THREAD

**Discussion by hillen** [rabbitmq/rabbitmq-server#11712] [Questions] [answered]
We have six different kubernetes clusters with three nodes each. In the last week, after updating to 3.13.4, we have been having issues where on a node image update the nodes have had the quorum queues go into a noproc state. It does not always happen, but happens more than half of the time. And it usually only happens for one of the nodes.
Once it gets into the state, it will log hundreds of these lines per second:
2024-07-12 03:47:43.972884+00:00 [warning] <0.6137.0> Mnesia->Khepri data copy: failed to setup Khepri projection for migration "rabbitmq_metadata", expect slower versions of `is_migration_finished()` and `handle_fallback()`
2024-07-12 03:47:43.972884+00:00 [warning] <0.6137.0> {error,noproc}

We have never enabled Khepri on any of our clusters.

We have been using rabbitmq in kubernetes for around 3.5 years and quorum queues for about a year and have not seen an issue like this before.

Once this happens, the only way that we have found to fix it is to delete the /opt/bitnami/rabbitmq/.rabbitmq/mnesia/rabbit@rabbitmq-2.rabbitmq-headless.rabbitmq.svc.cluster.local/quorum/rabbit@rabbitmq-2.rabbitmq-headless.rabbitmq.svc.cluster.local directory on the node with the noproc and to restart all of the nodes.

On the version update from 3.13.3 to 3.13.4 everything seemed fine. But the next time the nodes were restarted, the issue occurred. We did go back to 3.13.3 and on a restart it did the same thing. However, we were running 3.13.3 for over a month and some node restarts occurred and the issue did not occur. We did not see the issue until we moved to 3.13.4. Yet, going back to 3.13.3 and restarting the nodes it does occur. We cannot try 3.13.2 because the message_containers_deaths_v2 feature flag was enabled.

I am attaching the log of the failing node after the restart. I did not capture the logs on the previous shutdown, but I can probably pull them from our log storage if needed. We could also recreate the cluster with a version before 3.13.3 in one of our dev environments if necessary.
[rabbitmq-2-node.txt](https://github.com/user-attachments/files/16228543/rabbitmq-2-node.txt)

**Comment by dumbbell** [comment]
Hi!

The warning comes from the code that is responsible for selecting the correct database. Therefore it is used regardless of the active database. The situation is harmless and won't cause any issue. In this case, Mnesia is correctly selected.

The `noproc` reason is a bit weird as Khepri should run, even if not used. I will update the code to take this error into account and will rephrase the log message to be less scary. Perhaps I will also change the severity.

But in other words, you can ignore the warning.

**Comment by miklosfarkas** [comment]
I have a similar problem as well. Our system also logs a hundred lines per second. However, in our case, besides the warnings, there is also a phenomenon that after a few hours of operation, the problematic node drops out of the cluster. The node comes back after a stop and start. We are now planning to replace the problematic node with a new one.

**Reply by dumbbell** [reply]
Do you have logs from all nodes when that problematic node drops out of the cluster?

**Reply by kjnilsson** [reply]
@miklosfarkas logs would be really helpful here. Do you also have issues with quorum queues?

**Reply by miklosfarkas** [reply]
Hello, sorry I was offline!
Yes, I can share anything you need. I am downloading the log now. The problematic node has 29 GB, so please be patient

**Reply by miklosfarkas** [reply]
The meantime it turned out that the specific node was dropped from the cluster because the backup stopped it. 
The other nodes did not show the error because the virtual machines were not to large enough to take that long to backup
I changed the log level to error, so the warnings have disappeared and no more GB data is generated in the log

**Reply by miklosfarkas** [reply]
do you still need the data?

**Reply by kjnilsson** [reply]
Only if your issue isn't the same as the original reporter. Can you inspect the logs they provided and compare?

**Comment by kjnilsson** [comment]
Is there any chance you could share the quorum directory with me taken when one of these events happen?

**Reply by kjnilsson** [reply]
@hillen  when you say "Yet, going back to 3.13.3 and restarting the nodes it does occur." - this issue is the quorum queue noprocs? or the khepri logs?

**Reply by hillen** [reply]
Both, the quorum queue noprocs and the khepri logs occurs when going back to 3.13.3.
When going from 3.13.4 back to 3.14.3, the problem was actually worse in that in both clusters this was tried two of the three nodes went to noproc. And then the queues did not recover and had to be deleted and recreated after getting the rabbit nodes restarted without the issue.

To be clear, the problem does not happen on the change of version itself. It happens on the next restart. We are using helm.
helm upgrade or helm rollback that changes the version does not happen. But, the next kubectl rollout restart statefulset is when the error occurs.

After both the noproc and khepri logs no longer showed, I was able to do kubectl rollout restart statefulset calls without it failing. But, any time the nodes are rolled to update the kubernetes node, it happens a majority of the time.

**Reply by hillen** [reply]
I can try to make this happen in our one of our dev clusters manually after 5:00. What I will try to do is several kubectl rollout restart commands just to see if I can get it to occur. Since this seems to be a timing issue, I may have just not hit it on a restart.
If I cannot get it to happen with just the restarts.I will simulate Azure to a node update by increasing the node pool size and draining the old node. Wait a few minutes and drain the new node. And go through the three node pools. That should make it happen after a few tries.
Then I will share the quorum directory of the rabbitmq node with the noproc and one that is not showing the noproc.

Is there anything else that you want captured?

Once I reproduce and capture the information for 3.13.4, I will try to go back to 3.13.3 and capture logs and directories for that also.

**Reply by kjnilsson** [reply]
Thank you - if you could enable debug logging and share logs that would be helpful also.

So it only happened on 3.13 _after_ you had been on 3.14 then went back to 3.13?

RabbitMQ doesn't generally support downgrades but this would be an interesting data point anyway.

**Reply by kjnilsson** [reply]
What kind of load is running on this cluster? msg/sec / size of messages that kind of thing?

**Reply by hillen** [reply]
We went from 3.13.2 to 3.13.3 on June 1st. And from 3.13.3 to 3.13.4 on July 6th on our dev clusters. And the next week on our qa and prod clusters. The first that I noticed the problem was on July 8th. So, I did not see any issue with 3.13.3 until I downgraded and restarted. So, that may be a different issue caused by downgrading. During the time from going to 3.13.3 to 3.13.4 there were four helm chart upgrades, so there were four rollouts of the nodes with 3.13.3. I cannot tell you how many times Azure updated the node images during that time, but I would say at least twice.
So, it did appear to us that this was breakage with 3.13.4. However, when we rolled back to 3.13.3 and then restarted again we saw similar behavior with nodes being removed from the queue. But I do not have enough knowledge that looking at the logs I would have seen anything that I could recognize that it was not the same issue. It was a similar looking issue.

**Reply by kjnilsson** [reply]
Azure you say. What kind of storage are you using?

**Reply by hillen** [reply]
Looks like for the prod clusters they are running around 245 messages / second. The qa and dev clusters are around 5-10 messages / second. Nothing that I think you would consider high load. I am not sure how to get message sizes.

**Reply by hillen** [reply]
Looks like Azure is updating the nodes in west europe right now. I think that the rabbitmq-2 node in one of the clusters is going to have the issue. I should be able to get the quorum directories shortly.

**Reply by hillen** [reply]
Using a 10GB Azure disk

**Reply by kjnilsson** [reply]
yes 245 msgs/sec isn't a lot of load unless the messages are huge. We've had other users having some trouble with using Azure Files so it could be good to know what you are using there.

**Reply by hillen** [reply]
Our prod eu cluster is now in the failed state. Here is what I am seeing:
kubectl exec rabbitmq-0 -n rabbitmq -- rabbitmq-queues quorum_status SOLUTIONS.DOCUMENT_SUBMISSION.DOCUMENT_OVERLAY_SERVICE.DOCUMENT_PROCESSING_QUEUE_LISTENER.FOOTER.QUORUM
Status of quorum queue SOLUTIONS.DOCUMENT_SUBMISSION.DOCUMENT_OVERLAY_SERVICE.DOCUMENT_PROCESSING_QUEUE_LISTENER.FOOTER.QUORUM on node rabbit@rabbitmq-0.rabbitmq-headless.rabbitmq.svc.cluster.local ...
┌────────────────────────────────────────────────────────────────┬────────────┬────────────┬────────────────┬──────────────┬──────────────┬──────────────┬────────────────┬──────┬─────────────────┐
│ Node Name                                                      │ Raft State │ Membership │ Last Log Index │ Last Written │ Last Applied │ Commit Index │ Snapshot Index │ Term │ Machine Version │
├────────────────────────────────────────────────────────────────┼────────────┼────────────┼────────────────┼──────────────┼──────────────┼──────────────┼────────────────┼──────┼─────────────────┤
│ rabbit@rabbitmq-0.rabbitmq-headless.rabbitmq.svc.cluster.local │ follower   │ voter      │ 58             │ 58           │ 58           │ 58           │ 47             │ 10   │ 3               │
├────────────────────────────────────────────────────────────────┼────────────┼────────────┼────────────────┼──────────────┼──────────────┼──────────────┼────────────────┼──────┼─────────────────┤
│ rabbit@rabbitmq-1.rabbitmq-headless.rabbitmq.svc.cluster.local │ leader     │ voter      │ 58             │ 58           │ 58           │ 58           │ 47             │ 10   │ 3               │
├────────────────────────────────────────────────────────────────┼────────────┼────────────┼────────────────┼──────────────┼──────────────┼──────────────┼────────────────┼──────┼─────────────────┤
│ rabbit@rabbitmq-2.rabbitmq-headless.rabbitmq.svc.cluster.local │ noproc     │            │                │              │              │              │                │      │                 │
└────────────────────────────────────────────────────────────────┴────────────┴────────────┴────────────────┴──────────────┴──────────────┴──────────────┴────────────────┴──────┴─────────────────┘

**Reply by hillen** [reply]
I obviously did not have debugging logs on yet. I can try to pull the quorum directories from rabbitmq-2 and rabbitmq-0. Is there anything else that you want to see while it is in a failed state?

**Reply by kjnilsson** [reply]
if you could capture what logs you have that'd be great

**Reply by kjnilsson** [reply]
I have made my email public on my GH profile if you want to provide the quorum directory in a less public manner.

**Reply by kjnilsson** [reply]
Do your applications ever delete quorum queues? Or make changes to the members?

**Reply by hillen** [reply]
They should never delete them. All of our quorum queues should have all three rabbitmq nodes that are in the cluster as members.

**Reply by hillen** [reply]
We have three clusters in westeurope that Azure updated the kubernetes nodes today. Like last Friday when the eastus2 updates happened, two of the three clusters failed.

I am collecting the information and am going to send you the emails.

**Reply by kjnilsson** [reply]
Thanks, I will resume investigations in the morning.

**Reply by kjnilsson** [reply]
I couldn't find a 10GB Azure disk type here: https://azure.microsoft.com/en-us/pricing/details/managed-disks/

Could you point me to the exact type you are using?

**Reply by hillen** [reply]
It would be an S4 from that page. I had the size incorrect. It is 20GB. Here is the relevant information from the az show command:
  "diskIOPSReadWrite": 500,
  "diskMBpsReadWrite": 60,
  "diskSizeBytes": 21474836480,
  "diskSizeGB": 20,
  "sku": {
    "name": "Standard_LRS",
    "tier": "Standard"
  },
  "type": "Microsoft.Compute/disks",

**Reply by hillen** [reply]
The emails with the quorum directory zip file attachments are getting rejected because of the 20MB size limit when trying to send the emails. Do you have a Google Cloud account that you could temporarily make public or give me temporary credentials to?

**Reply by hillen** [reply]
I believe that I was able to send you the files for the dev environment that failed today also. The files for that environment are about 20x smaller and should be a total of about 5MB. That should make it through the Google size limit for attachments.

If you want the ones for the other environment that failed today. Then we will need to figure out how to get those files somewhere that you can pull them down. I do not have the logs for that environment. With the lines that were being generated hundreds of times a second, it rotated the logs.

**Reply by kjnilsson** [reply]
thanks for sending me the files. They validated what I suspected may be the cause last night and I am pretty sure I have a fix for it now in https://github.com/rabbitmq/ra/pull/456.

**Reply by kjnilsson** [reply]
Ok I think we have a fix for it. if you have any capacity to test it there is an image here https://hub.docker.com/layers/pivotalrabbitmq/rabbitmq/rabbitmq-main-ra-82c4cbc-26.2.5.2/images/sha256-3b65c1264e6f713b86b4359744c6b73994c5c55da6bf2fea953f03ed2c94f04e

this is run off RabbitMQ main so there is no downgrade path for any environment you use with this.

**Reply by hillen** [reply]
I am not sure that I can test this container. I am using the bitnami helm chart and the directories specified in the statefulset match what is expected in the container. There are some scripts that are run when the container is built to ensure the expected directories are there. Unless you have experienced this in the past know that I can just swap out the container. I could build the container using the bitnami Dockerfile if I had a rabbitmq build tarball in a format that matches this download: curl -SsLf "https://downloads.bitnami.com/files/stacksmith/rabbitmq-3.13.4-0-linux-amd64-debian-12.tar.gz" -O

**Reply by hillen** [reply]
Do you have any idea when this fix may be released in a new version of rabbitmq?
Do you know of anything that can be done to patch a failing node other than delete the quorum directory? So far, we have only seen one node get into the noproc state each time that it failed. And other than some queues backing up because any service connected to that node does not get messages, the effect has been fairly minimal. But if two nodes would go out, then I believe we would start seeing the environment having noticeable issues and perhaps messages would be lost if the quorum directory had to be deleted from two nodes.

**Reply by kjnilsson** [reply]
We need to validate the fix - right now I am struggling to reproduce the issue inside RabbitMQ. Once we have validated I think the fix will go out relatively soon but can't give any definitive dates.

**Reply by kjnilsson** [reply]
How many cpu cores do your RabbitMQ servers have?

**Reply by kjnilsson** [reply, marked as answer]
Here is a PR that addresses this issue https://github.com/rabbitmq/rabbitmq-server/pull/11758

**Reply by michaelklishin** [reply]
https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.13.5 is out. Our team cannot comment on when Bitnami images might adopt this version. You are welcome to politely ask them to do it ASAP.

**Comment by michaelklishin** [comment]
https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.13.5 is out. Our team cannot comment on when Bitnami images might adopt this version. You are welcome to politely ask them to do it ASAP.

**Reply by hillen** [reply]
The servers running rabbitmq are the following: Standard_D4as_v5
We are running three kubernetes nodes, one in each availability zone, and one rabbitmq node on each kubernetes node.
Other than the daemonsets, this is all that is running on the kubernetes nodes.
So, four cores on the VMs with rabbitmq having the ability to use most of them.

**Reply by hillen** [reply]
Thanks for getting the release out quickly.
Looks like bitnami built the release around 4:00am EDT Saturday:
31407225 Jul 20 03:59 rabbitmq-3.13.5-0-linux-amd64-debian-12.tar.gz
However, the container bot has not changed the container Dockerfile and released it. From the history, it looks like a lot of this is done around 6:30am but I logged an issue just to make sure something is not stuck in the process.
https://github.com/bitnami/containers/issues/69428
